### PR TITLE
CA Release

### PR DESCRIPTION
**Pull request by taehyunlim**
---
CA Release
---

### Description
This PR includes changes required for CA support. 

### Summary of changes
* Store view level Affirm merchant credential configuration for US and CA regions:
Screen Shot 2022-08-16 at 11 54 52 PM
* Affirm checkout object in global format
* Global API and script URLs
* CAD currency support
* Locale based checkout flow:
Screen Shot 2022-08-16 at 11 58 20 PM

### Additional information
* Notable changes include relaxing strict mode in `_affirm_config` loading scripts due to scope issues
* Country code is currently tied to currency (USD->USA; CAD->CAN) 
* Adds CSP policy exceptions to resolve security warning messages
* Post-checkout API request validator uses payment data for amount validation (as opposed to subject attributes) in order to support non-base currency checkouts

### CODE REVIEW

File: Gateway/Validator/Client/PaymentActionsValidator.php [hunk 1]
```diff
@@ -33,6 +33,22 @@ public function validate(array $validationSubject)
     {
         $response = SubjectReader::readResponse($validationSubject);
         $amount = SubjectReader::readAmount($validationSubject);
+
```
[the remark this answers is not in the document]
Comment (kkong87) [reply Gateway/Validator/Client/PaymentActionsValidator.php:36]: Can we clean this so there is an else statement

**Review by kkong87 — COMMENTED** [review]
added some comments

File: Gateway/Request/CaptureRequest.php [hunk 2]
```diff
@@ -46,12 +46,13 @@ public function build(array $buildSubject)
...
         $transactionId = $payment->getAdditionalInformation(self::TRANSACTION_ID) ?:
             $payment->getAdditionalInformation(self::CHARGE_ID);
+        $countryCode = $payment->getAdditionalInformation(self::COUNTRY_CODE)?: 'USA';
         $order = $payment->getOrder();
         $storeId = isset($order) ? $order->getStoreId() : $this->_storeManager->getStore()->getId();
         if (!$storeId) {
             $storeId = null;
         }
-        if ($this->affirmPaymentConfig->getPartialCapture()) {
+        if ($this->affirmPaymentConfig->getPartialCapture($countryCode)) {
```
[the remark this answers is not in the document]
Comment (taehyunlim) [reply Gateway/Request/CaptureRequest.php:55]: `/Model/Plugin/Payment/CanCapturePartial.php` will use the country code to disable partial capture in CA

File: Gateway/Http/AbstractTransferFactory.php [hunk 3]
```diff
@@ -82,16 +92,17 @@ public function __construct(
...
+    protected function getPublicApiKey($storeId, $country_code)
     {
+        $country_suffix = $this->getApiKeyNameByCountry($country_code);
         if(!empty($storeId)){
-            return $this->config->getValue('mode', $storeId) == 'sandbox'
-                ? $this->config->getValue('public_api_key_sandbox', $storeId)
-                : $this->config->getValue('public_api_key_production', $storeId);
+            return $this->config->getValue(self::MODE, $storeId) == self::SANDBOX
+                ? $this->config->getValue('public_api_key_sandbox' . $country_suffix, $storeId)
+                : $this->config->getValue('public_api_key_production' . $country_suffix, $storeId);
```
[the remark this answers is not in the document]
Comment (taehyunlim) [reply Gateway/Http/AbstractTransferFactory.php:101]: Yep, it's store as a config eav attribute `public_api_key_sandbox_ca` and `public_api_key_production_ca`

File: Gateway/Validator/Client/PaymentActionsValidator.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (taehyunlim) [reply Gateway/Validator/Client/PaymentActionsValidator.php:36]: updated with else statement

File: Gateway/Http/AbstractTransferFactory.php [hunk 4]
```diff
@@ -124,4 +136,17 @@ protected function getStoreId()
...
+
+    /**
+     * Map country code to API key config name
+     *
+     * @return string
+     */
+    protected function getApiKeyNameByCountry($country_code)
+    {
+        if ($country_code == 'CAN') {
+            return self::SUFFIX_CANADA;
```
[the remark this answers is not in the document]
Comment (taehyunlim) [reply Gateway/Http/AbstractTransferFactory.php]: updated

File: Gateway/Http/AbstractTransferFactory.php [hunk 5]
```diff
@@ -138,15 +140,21 @@ protected function getStoreId()
...
      *
      * @return string
      */
     protected function getApiKeyNameByCountry($country_code)
     {
-        if ($country_code == 'CAN') {
-            return self::SUFFIX_CANADA;
+        $_suffix = '';
+        $countryCodeToSuffix = array(
+            'CAN' => self::SUFFIX_CANADA,
```
Comment (nandadubey) [inline Gateway/Http/AbstractTransferFactory.php:154]: Can you use COUNTRY_CODE_CAN and COUNTRY_CODE_USA instead

File: Gateway/Request/CaptureRequest.php [hunk 6]
```diff
@@ -46,12 +46,13 @@ public function build(array $buildSubject)
         $payment = $paymentDataObject->getPayment();
         $transactionId = $payment->getAdditionalInformation(self::TRANSACTION_ID) ?:
             $payment->getAdditionalInformation(self::CHARGE_ID);
+        $countryCode = $payment->getAdditionalInformation(self::COUNTRY_CODE)?: 'USA';
```
Comment (nandadubey) [inline Gateway/Request/CaptureRequest.php:49]: can we use default_country_code enum instead of USA

File: Gateway/Request/RefundRequest.php [hunk 7]
```diff
@@ -46,7 +46,9 @@ public function build(array $buildSubject)
         $payment = $paymentDataObject->getPayment();
         $transactionId = $payment->getAdditionalInformation(self::TRANSACTION_ID) ?:
             $payment->getAdditionalInformation(self::CHARGE_ID);
-        $amountInCents = Util::formatToCents($buildSubject['amount']);
+        $countryCode = $payment->getAdditionalInformation(self::COUNTRY_CODE) ?: 'USA';
+        $creditMemoAmount = $payment->getCreditmemo()->getGrandTotal();
+        $amountInCents = Util::formatToCents($creditMemoAmount);
```
Comment (nandadubey) [inline Gateway/Request/RefundRequest.php:51]: will this support a partial amount refund?

File: Gateway/Validator/Client/PaymentActionsValidator.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (nandadubey) [reply Gateway/Validator/Client/PaymentActionsValidator.php:36]: I have the same question, can you add a comment in the code itself what is this logic for?

File: Helper/Payment.php [hunk 8]
```diff
@@ -311,7 +311,7 @@ public function getConfigurableProductBackordersOptions(Product $product = null)
     public function validateVirtual()
     {
         if ($this->quote->getIsVirtual() && !$this->quote->getCustomerIsGuest()) {
-            $countryId = self::VALIDATE_COUNTRY;
+            $countryId = '';
```
Comment (nandadubey) [inline Helper/Payment.php:314]: why are we not adding default value instead keeping it empty to start with? what happens if it never goes in for loop?

File: Model/Config.php [hunk 9]
```diff
@@ -654,12 +692,32 @@ public function getCheckoutFlowType()
     }
 
     /**
-     * Get partial capture
+     * Get partial capture (US only)
      *
      * @return bool
      */
-    public function getPartialCapture()
+    public function getPartialCapture($countryCode = 'USA')
```
Comment (nandadubey) [inline Model/Config.php:699]: Please use global enum/constants instead of hard-coded value

File: view/frontend/web/js/model/aslowas.js [hunk 10]
```diff
@@ -32,14 +31,15 @@ define([
...
             process: function(price, options) {
                 self = this;
-                var formatted, priceInt, optionsPrice;
-                formatted = Number(price.replace(/[^0-9\.]+/g,""));
+                var formatted, priceInt, optionsPrice, priceStr;
 
-                if (options.currency_rate) {
-                    formatted = formatted / options.currency_rate;
-                    formatted = formatted.toFixed(2);
+                priceStr = price.replace(/^(-)|[^0-9.,]+/g, '$1').trim();
```
Comment (nandadubey) [inline view/frontend/web/js/model/aslowas.js:36]: move the logic in the helper function and ideally if possible introduce currency for parsing logic to make it more future-proof (future when we deal with a currency that is not multiplied by 100 for the lowest unit, not needed right now though)

File: Gateway/Request/RefundRequest.php (hunk 7 above)
Comment (nandadubey) [inline Gateway/Request/RefundRequest.php:51]: if possible introduce currency in formatToCents to make it more future-proof (future when we deal with a currency that is not multiplied by 100 for the lowest unit, not needed right now though)

File: Gateway/Validator/Client/PaymentActionsValidator.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (taehyunlim) [reply Gateway/Validator/Client/PaymentActionsValidator.php:36]: added additional comments

File: Gateway/Request/CaptureRequest.php (hunk 6 above)
Comment (taehyunlim) [reply Gateway/Request/CaptureRequest.php:49]: added an enum and changed to: `$countryCode = $payment->getAdditionalInformation(self::COUNTRY_CODE) ?: self::DEFAULT_COUNTRY_CODE;`

File: Gateway/Request/RefundRequest.php (hunk 7 above)
Comment (taehyunlim) [reply Gateway/Request/RefundRequest.php:51]: Yes

File: Helper/Payment.php (hunk 8 above)
Comment (taehyunlim) [reply Helper/Payment.php:314]: Yep, it is better to assign to a default value. Added `DEFAULT_REGION`

File: Model/Config.php (hunk 9 above)
Comment (taehyunlim) [reply Model/Config.php:710]: replaced hardcoded value throughout the class

File: Gateway/Request/RefundRequest.php (hunk 7 above)
Comment (taehyunlim) [reply Gateway/Request/RefundRequest.php:51]: [This commit](https://github.com/Affirm/Magento2_Affirm/pull/81/commits/4b80b2555afb3d2f4a1b1c72c3a4195c65eb729a) introduces currency parameters to the helper functions.

File: view/frontend/web/js/model/aslowas.js (hunk 10 above)
Comment (taehyunlim) [reply view/frontend/web/js/model/aslowas.js]: (Same as below) [This commit](https://github.com/Affirm/Magento2_Affirm/pull/81/commits/4b80b2555afb3d2f4a1b1c72c3a4195c65eb729a) introduces currency parameters to the helper functions.

File: Gateway/Helper/Util.php [hunk 11]
```diff
@@ -52,7 +62,7 @@ public function __construct(
      * @param int $amount
      * @return int
      */
-    public static function formatToCents($amount = 0)
+    public static function formatToCents($amount = 0, $currencyCode = self::CURRENCY_CODE_USD)
```
Comment (kkong87) [inline Gateway/Helper/Util.php:65]: Can we update the comment to include currency code.

Comment (kkong87) [reply Gateway/Helper/Util.php:65]: is there a reason why we are passing currency code, that doesn't look like its being used

File: Model/Config.php [hunk 12]
```diff
@@ -207,11 +229,7 @@ public function isCurrentStoreCurrencyUSD()
     {
         $currentCurrency = $this->storeManager->getStore()
             ->getCurrentCurrencyCode();
-        $isUSD = true;
-        if ($currentCurrency != self::CURRENCY_CODE) {
-            $isUSD = false;
-        }
-        return $isUSD;
+        return $currentCurrency == self::CURRENCY_CODE_USD;
```
Comment (kkong87) [inline Model/Config.php:232]: This just returns USD, do we need to add other currencies?

**Review by kkong87 — COMMENTED** [review]
added some comments

File: Model/Config.php (hunk 12 above)
Comment (taehyunlim) [reply Model/Config.php:232]: This function checks if the currency is USD and returns a bool

**Review by nandadubey — APPROVED** [review]
LGTM, considering all UAT tests are passing and all US-specific features like Slit Capture and Telesales gracefully errors out or restricts it's availability

File: Gateway/Helper/Util.php (hunk 11 above)
Comment (taehyunlim) [reply Gateway/Helper/Util.php]: reverted this change

**Pull request merged**